illumos-utils/src/opte/port_manager.rs

```diff
@@ -108,13 +111,213 @@ impl PortManagerInner {
+        let has_ipv6_stack = ip_config.ipv6_addr().is_some();
+
+        // First, attempt to convert the IPv4 configuration.
+        //
+        // The important part here is ensuring that we have an external and
+        // VPC-private configuration for each IP stack. That is, if we have an
+        // IPv4 external Ephemeral IP, we also have an IPv4 VPC-private address.
+        let v4 = match ip_config.ipv4_config() {
+            None => {
+                params.ensure_no_external_ipv4_addresses()?;
+                None
+            }
+            Some(ipv4_config) => {
+                let gateway_ip = ipv4_config.subnet().first_host().into();
+                let vpc_subnet =
+                    Ipv4Cidr::from(Ipv4Network::from(*ipv4_config.subnet()));
+                let private_ip = (*ipv4_config.ip()).into();
+                let external_ips = build_external_ipv4_config(
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    has_ipv4_stack,
+                )?;
+                Some(Ipv4Cfg {
+                    vpc_subnet,
+                    private_ip,
+                    gateway_ip,
+                    external_ips,
+                })
+            }
+        };
+
+        // Now the same conversion, but for IPv6. Again, we need to ensure that,
+        // if we have an external IPv6 address, we also have a VPC-private IPv6
+        // address to translate that into.
+        let v6 = match ip_config.ipv6_config() {
+            None => {
+                params.ensure_no_external_ipv6_addresses()?;
+                None
+            }
+            Some(ipv6_config) => {
+                let gateway_ip = ipv6_config
+                    .subnet()
+                    .iter()
+                    .nth(1)
+                    .expect("must have at least 2 addresses")
+                    .into();
+                let vpc_subnet =
+                    Ipv6Cidr::from(Ipv6Network::from(*ipv6_config.subnet()));
+                let private_ip = (*ipv6_config.ip()).into();
+                let external_ips = build_external_ipv6_config(
+                    params.source_nat.as_ref(),
+                    params.ephemeral_ip.as_ref(),
+                    &params.floating_ips,
+                    has_ipv6_stack,
+                )?;
+                Some(Ipv6Cfg {
+                    vpc_subnet,
+                    private_ip,
+                    gateway_ip,
+                    external_ips,
+                })
+            }
+        };
+
+        // Now build the full configuration, either single- or dual-stack.
+        let cfg = match (v4, v6) {
+            (Some(ipv4), Some(ipv6)) => IpCfg::DualStack { ipv4, ipv6 },
+            (Some(v4), None) => IpCfg::Ipv4(v4),
+            (None, Some(v6)) => IpCfg::Ipv6(v6),
+            (None, None) => unreachable!(),
```

Comment:
Worth a comment that `v6` and `v4` come from a `PrivateIpConfig` and it's not possible for an instance of that enum to not yield either a v4 or v6 address. Alternatively it seems like the logic in this function could just match on `PrivateIpConfig` directly and avoid the unreachable match branch all together.

Reply:
I've reworked this to use a match like you suggested, thanks!